### PR TITLE
Add donate 1

### LINKED ISSUE #21: Add a donation page
**Is your feature request related to a problem? Please describe.**
Currently there is no way to donate to SithCult.

**Describe the solution you'd like**
A new way to donate should be added in form of a new `contribution` page.

### PR DESCRIPTION
**Pull request by Theralun**
**Checklist:**

 
 - [ ] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [ ] Have you updated or added documentation for the change?
 - [x] Have you tested your changes with successful results?

**Type of Changes:**

 
 - [x] Bug fix (non-breaking change which fixes an issue)
 - [x] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change)

**What is the current behavior? (link to any open issues here)**
- Currently you cannot donate to SithCult. #21 
- Currently the sign up of users is not really stable.
- Currently there are some files which are not up to our standards.

**What is the new behavior (if this is a feature change)?**
- It is now possible to donate to SithCult via Paypal. Closes #21 
- The sign up has been improved and should now be more stable.
- Files are now up to our 2020/06 standards.

**Other information**:
- Minor update (1.4.0).

### CODE REVIEW

File: src/components/molecules/ReceivedUser/index.jsx [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+class ReceivedUser extends React.Component {
+  handlePopoverChange = (open) => {
+    if (!open) {
+      this.props.clearUser();
+    }
+  };
+
+  render() {
+    const { receivedUser, author } = this.props;
```
Comment (LucaAllmaier) [inline src/components/molecules/ReceivedUser/index.jsx:30]: add new line.

Comment (Aichnerc) [inline src/components/molecules/ReceivedUser/index.jsx:30]: Add a new line.

File: src/store/actions/postActions.js [hunk 2]
```diff
@@ -156,6 +156,36 @@ export const loadPosts = (amount) => {
...
+        }
+
+        // Create post
+        firestore.collection('posts').doc(postId).set({
+            comments: localComments,
+        }, { merge: true }).then(() => {
+            dispatch({ type: 'COMMENT_SUCCESS', postId });
+            return 
+        }).catch((err) => {
+            dispatch({ type: 'COMMENT_ERROR', err });
```
Comment (LucaAllmaier) [inline src/store/actions/postActions.js:184]: please use " " instead of ' '

File: src/components/molecules/index.js [hunk 3]
```diff
@@ -1,10 +1,16 @@
...
+import Navbar from "./Navbar";
+import Footer from "./Footer";
+import ReceivedUser from "./ReceivedUser";
 
-export { Footer, Navbar, CookieModal };
+export {
+  CookieModal,
+  Footer,
+  Navbar,
+  ReceivedUser,
```
Comment (Aichnerc) [inline src/components/molecules/index.js:12]: Export in the same order as imported.
```suggestion
  CookieModal,
  Navbar,
  Footer,
  ReceivedUser,
```

File: src/components/organisms/RegisterForm/index.jsx [hunk 4]
```diff
@@ -49,7 +49,6 @@ import IMGlogo from "../../../assets/images/logo_white.png";
 
 // Country list
 const countries = countryList().getData();
```
Comment (Aichnerc) [inline src/components/organisms/RegisterForm/index.jsx:51]: Add new line.

File: src/store/actions/postActions.js [hunk 5]
```diff
@@ -156,6 +156,36 @@ export const loadPosts = (amount) => {
   };
 };
 
+export const commentPost = (postId, comment, previousComments) => {
+    return (dispatch, getState, {getFirebase, getFirestore}) => {
+        const firebase = getFirebase();
+        const firestore = getFirestore();
+
+        let localComments = previousComments;
```
Comment (LucaAllmaier) [inline src/store/actions/postActions.js:164]: add new line.

File: src/components/organisms/RegisterForm/index.jsx [hunk 6]
```diff
@@ -73,17 +72,17 @@ class HomePage extends React.Component {
...
 
   componentDidMount = () => {
-    this._getIPData();
+    this._fetchAllCountries();
   };
 
   submitHandler = (event) => {
     event.preventDefault();
 
     let result = this._signUserUp();
```
Comment (Aichnerc) [inline src/components/organisms/RegisterForm/index.jsx:81]: Add new line.

File: src/components/pages/DonatePage/index.jsx [hunk 7]
```diff
@@ -0,0 +1,599 @@
+//> React
+// Contains all the functionality necessary to define React components
+import React from "react";
+
+//> Additional
+import { PayPalButton } from "react-paypal-button-v2";
+import moment from "moment";
+import Confetti from "react-confetti";
```
Comment (Aichnerc) [inline src/components/pages/DonatePage/index.jsx:8]: Please explain what those packages do.

File: src/components/pages/DonatePage/index.jsx [hunk 8]
```diff
@@ -0,0 +1,599 @@
...
+// Connect
+import { connect } from "react-redux";
+// Actions
+import {
+  updateBadgesDonate,
+  writeDonation,
+  getDonations,
+} from "../../../store/actions/userActions";
+
+import { writeDonationMessage } from "../../../store/actions/donMsgActions";
```
Comment (Aichnerc) [inline src/components/pages/DonatePage/index.jsx:39]: Remove the line above.

File: src/components/pages/HomePage/index.jsx [hunk 9]
```diff
@@ -7,13 +7,13 @@ import React from "react";
...
-    );
-  }
+	render() {
+		return (
+			<>
+				<RegisterForm />
+			</>
+		);
+	}
 }
```
Comment (Aichnerc) [inline src/components/pages/HomePage/index.jsx:17]: The indentations do not look right. Try using prettier.

File: src/store/actions/authActions.js [hunk 10]
```diff
@@ -99,28 +99,38 @@ export const signUp = (newUser) => {
...
+                })
+                .then(function () {
+                  console.log("User successfully written!");
+                })
+                .catch(function (error) {
+                  console.error("Error creating user: ", error);
+                });
             })
             .then(() => {
+              console.log("Success Signup");
```
Comment (Aichnerc) [inline src/store/actions/authActions.js:132]: No need for that console.log. Such things should be handeled in the reducers.

File: src/store/actions/postActions.js (hunk 2 above)
Comment (Aichnerc) [reply src/store/actions/postActions.js:184]: He should use Prettier on this file, yes.

File: src/store/actions/userActions.js [hunk 11]
```diff
@@ -23,6 +23,116 @@ export const getUser = (uid) => {
...
+
+    firestore
+      .collection("donations")
+      .get()
+      .then((querySnapshot) => {
+        let results = [];
+
+        !querySnapshot.empty &&
+          querySnapshot.forEach((doc) => {
+            let data = { ...doc.data(), id: doc.id };
```
Comment (Aichnerc) [inline src/store/actions/userActions.js:39]: Add new line.

File: src/components/pages/DonatePage/index.jsx [hunk 12]
```diff
@@ -0,0 +1,599 @@
...
+                  Humanity. All members of Sith Cult and beyond who long for a
+                  Sith Imperial society.
+                </p>
+                <p>
+                  <strong>What will the funds be used for?</strong>
+                  <br />
+                  We will use our gained ressources to
+                  <br />- further promote SithCult
+                  <br />- provide international SithCult clusters with funds to
+                  create banners and flyer
```
Comment (LucaAllmaier) [inline src/components/pages/DonatePage/index.jsx:438]: flyers :D

File: src/store/actions/userActions.js [hunk 13]
```diff
@@ -23,6 +23,116 @@ export const getUser = (uid) => {
...
+  details,
+  credits,
+  reputation,
+  sith_name
+) => {
+  return (dispatch, getState, { getFirebase, getFirestore }) => {
+    const firebase = getFirebase();
+    const firestore = getFirestore();
+
+    const uid = firebase.auth().currentUser.uid;
```
Comment (Aichnerc) [inline src/store/actions/userActions.js:67]: Add comment.
```suggestion
    // Get the unique ID of the current user
    const uid = firebase.auth().currentUser.uid;
```

File: src/store/actions/userActions.js [hunk 14]
```diff
@@ -23,6 +23,116 @@ export const getUser = (uid) => {
...
+      Array.prototype.push.apply(badges, newBadges);
+    }
+
+    // Get current timestamp
+    const timestamp = new Date().getTime();
+
+    // Check amount
+    const amount = details?.purchase_units[0]?.amount?.value;
+
+    const newCredits = Math.round(credits + amount * 9);
```
Comment (Aichnerc) [inline src/store/actions/userActions.js:87]: Add comment.
```suggestion
    // Add Imperial Credits and reputation to the Sith user
    const newCredits = Math.round(credits + amount * 9);
```

File: src/components/pages/DonatePage/index.jsx [hunk 15]
```diff
@@ -0,0 +1,599 @@
...
+                  </strong>
+                  <br />
+                  When logged in, you will receive a "Founder" and "Phase 1"
+                  badge in our social network "Holobook", along with{" "}
+                  {Math.round(this.state.selectedAmount / 2)} reputation and{" "}
+                  {Math.round(this.state.selectedAmount * 9)} virtual Imperial
+                  credits on Holobook. This can be used to gain higher ranks and
+                  make others feel your presence and influence. Every supporter
+                  will receive a personal certificate which will reflect your
+                  passion for Sith Cult and the Sith Empire.
```
Comment (LucaAllmaier) [inline src/components/pages/DonatePage/index.jsx:471]: ... which will reflect _their_ passion ...

File: src/store/reducers/rootReducer.js [hunk 16]
```diff
@@ -1,29 +1,31 @@
...
-    auth: authReducer,
-    post: postReducer,
-    user: userReducer,
-    firestore: firestoreReducer,
-    firebase: firebaseReducer // Authentication
-})
+  auth: authReducer,
+  post: postReducer,
+  user: userReducer,
+  donMsg: donMsgReducer,
```
Comment (Aichnerc) [inline src/store/reducers/rootReducer.js:21]: Please add comments of what each actions and their reducers do.
```suggestion
  auth: authReducer, // User authentication
  post: postReducer, // Post management
  user: userReducer, // User related functionality
  donMsg: donMsgReducer, // Donation messages
```

**Pull request merged**